src/AddThreepid.js

```diff
@@ -171,10 +173,37 @@ export default class AddThreepid {
                         id_access_token: identityAccessToken,
                     });
                 } else {
-                    await MatrixClientPeg.get().addThreePidOnly({
-                        sid: this.sessionId,
-                        client_secret: this.clientSecret,
-                    });
+                    try {
+                        await this._makeAddThreepidOnlyRequest();
+
+                        // The spec has always required this to use UI auth but synapse briefly
+                        // implemented it without, so this may just succeed and that's OK.
+                        return;
+                    } catch (e) {
+                        if (e.httpStatus !== 401 || !e.data || !e.data.flows) {
+                            // doesn't look like an interactive-auth failure
+                            throw e;
+                        }
+
+                        // pop up an interactive auth dialog
+                        const InteractiveAuthDialog = sdk.getComponent("dialogs.InteractiveAuthDialog");
+
+                        return new Promise((resolve, reject) => {
+                            Modal.createTrackedDialog('Add Email', '', InteractiveAuthDialog, {
```

Comment:
I believe `Modal` now supports promises internally:

```suggestion
                            const { finished } = Modal.createTrackedDialog('Add Email', '', InteractiveAuthDialog, {
```

Then you can return `finished` instead of wrapping your own.

Reply:
oh, hah - that's much nicer.